loom/engine/bindings/loom/lmModestMaps.cpp

```diff
@@ -0,0 +1,311 @@
+        float zoomedRow = row * zoomFactor;
+                    
+        worldMatrix->transformCoordInternal(zoomedColumn, zoomedRow, &lastCoordinateX, &lastCoordinateY);
+
+        //transform into correct space if necessary
+        if ((context != NULL) && (context != object))
+        {
+            localToGlobal((DisplayObject *)(object->parent), &lastCoordinateX, &lastCoordinateY);
+            globalToLocal(context, &lastCoordinateX, &lastCoordinateY);
+        }
+    } 
+
+
+    static void setGridCoordinates(Matrix *invMatrix, float mapWidth, float mapHeight, float mapScale)
+    {
+        const float LN2 = 0.6931471805599453f;
+        gridZoom = log(mapScale) / LN2;
+        invMatrix->transformCoordInternal(0.0f, 0.0f, &gridTLy, &gridTLx);
+        invMatrix->transformCoordInternal(mapWidth, mapHeight, &gridBRy, &gridBRx);
+        invMatrix->transformCoordInternal(mapWidth, 0.0f, &gridTRy, &gridTRx);
+        invMatrix->transformCoordInternal(0.0f, mapHeight, &gridBLy, &gridBLx);
+        invMatrix->transformCoordInternal(mapWidth * 0.5f, mapHeight * 0.5f, &gridCy, &gridCx);
+    }
+
+
+    static Matrix *getGridInverseMatrix(Matrix *worldMatrix, float tileWidth, float tileHeight, float mapScale)
+    {
+        Matrix *invMatrix = new Matrix();
+        invMatrix->invertOther(worldMatrix);
+        invMatrix->scale(mapScale / tileWidth, mapScale / tileHeight);
+        return invMatrix;
+    }
+
+ 
+    static const char *getMSProviderZoomString(float col, float row, int zoom)
+    {
+        LOOM_PROFILE_SCOPE(mmZoom);
+        // we don't wrap rows here because the map/grid should be enforcing outerLimits :)
+
+
+        LOOM_PROFILE_START(mmZoomPre);
+
+        float zoomExp = pow(2.0f, zoom);
+        float wrappedColumn = fmod(col, zoomExp);
+        while (wrappedColumn < 0)
+        {
+            wrappedColumn += zoomExp;
+        }
+        col = wrappedColumn;
+        
+        // convert row + col to zoom string
+        // padded with zeroes so we end up with zoom digits after slicing:
+        convertToBinary((int) row, _rowBinaryString);
+        convertToBinary((int) col, _colBinaryString);
+
+        LOOM_PROFILE_END(mmZoomPre);
+
+        LOOM_PROFILE_START(mmZoomLoop);
+
+        // generate zoom string
+        int rowOffset = strlen(_rowBinaryString) - zoom;
+        int colOffset = strlen(_colBinaryString) - zoom;
+        char *zoomString = (char*)lmAlloc(NULL, sizeof(char) * (zoom + 1));
+        for(int i = 0; i < zoom; i++) 
+        {
+            //proces the row and col bits to build up the zoom string; values of 0,1,2,3
+            char value = _colBinaryString[i + colOffset];
+            if(_rowBinaryString[i + rowOffset] == '1')
+            {
+                value = (value == '1') ? '3' : '2';
+            }
+            zoomString[i] = value;
+        }
+        zoomString[zoom] = '\0';
+
+        LOOM_PROFILE_END(mmZoomLoop);
+        return (const char *)zoomString; 
```

Comment:
IIRC it's copy on return with the loomscript bridge. So you can just make it a static buffer and return and you avoid malloc activity and also potential leaks.